File: backend_fastapi/app/core/policy.py
from __future__ import annotations
import uuid
import re
from datetime import datetime, timezone
from typing import Any


def make_id(prefix: str = "id") -> str:
    return f"{prefix}_{uuid.uuid4().hex[:20]}"


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _normalize_scope(s: str = "") -> str:
    return str(s).strip().lower()


def scope_allows(granted: str, requested: str) -> bool:
    g = _normalize_scope(granted)
    r = _normalize_scope(requested)
    if g == "*" or g == r:
        return True
    if g.endswith(":*"):
        return r.startswith(g[:-1])
    return False


def evaluate_permit(
    *,
    permit: dict[str, Any] | None = None,
    request: dict[str, Any] | None = None,
    now: datetime | None = None,
) -> dict[str, Any]:
    now = now or datetime.now(timezone.utc)
    request = request or {}
    if not permit:
        return {"decision": "blocked", "code": "permit_not_found", "reason": "No permit was supplied"}
    status = permit.get("status", "")
    if status != "active":
        return {"decision": "blocked", "code": f"permit_{status}", "reason": f"Permit is {status}"}
    expires_at = permit.get("expiresAt") or permit.get("expires_at")
    if expires_at:
        if isinstance(expires_at, str):
            expires_dt = datetime.fromisoformat(expires_at.replace("Z", "+00:00")).replace(tzinfo=None)
        else:
            expires_dt = datetime.fromtimestamp(expires_at, tz=timezone.utc).replace(tzinfo=None)
        if expires_dt <= now.replace(tzinfo=None):
            return {"decision": "blocked", "code": "permit_expired", "reason": "Permit has expired"}
    agent_id = permit.get("agentId") or permit.get("agent_id")
    req_agent_id = request.get("agentId") or request.get("agent_id")
    if agent_id and req_agent_id and agent_id != req_agent_id:
        return {"decision": "blocked", "code": "agent_mismatch", "reason": "Permit belongs to another agent"}
    scopes = permit.get("scopes", []) or permit.get("scopes", [])
    req_scope = request.get("scope", "")
    allowed = any(scope_allows(s, req_scope) for s in scopes)
    if not allowed:
        return {"decision": "blocked", "code": "scope_denied", "reason": f"Scope {req_scope} is not granted"}
    amount = float(request.get("amount", 0))
    if amount < 0:
        return {"decision": "blocked", "code": "invalid_amount", "reason": "Amount must be non-negative"}
    max_per_action = permit.get("maxPerAction") or permit.get("max_per_action")
    if max_per_action is not None and amount > float(max_per_action):
        return {"decision": "escalated", "code": "per_action_limit", "reason": f"Amount exceeds per-action limit of {max_per_action}"}
    require_human_above = permit.get("requireHumanAbove") or permit.get("require_human_above")
    if require_human_above is not None and amount > float(require_human_above):
        return {"decision": "escalated", "code": "human_threshold", "reason": f"Amount requires human approval above {require_human_above}"}
    allowed_targets = permit.get("allowedTargets", []) or permit.get("allowed_targets", [])
    req_target = request.get("target")
    if allowed_targets and req_target and req_target not in allowed_targets:
        return {"decision": "blocked", "code": "target_denied", "reason": "Target is not on the permit allowlist"}
    conditions = permit.get("conditions", []) or permit.get("conditions", [])
    for cond in conditions:
        field_value = request.get(cond["field"]) or (request.get("context") or {}).get(cond["field"])
        if field_value is None:
            continue
        passes = False
        op = cond.get("operator", "")
        val = cond.get("value")
        if op == "eq":
            passes = str(field_value) == str(val)
        elif op == "neq":
            passes = str(field_value) != str(val)
        elif op == "lte":
            passes = float(field_value) <= float(val)
        elif op == "lt":
            passes = float(field_value) < float(val)
        elif op == "gte":
            passes = float(field_value) >= float(val)
        elif op == "gt":
            passes = float(field_value) > float(val)
        elif op == "in":
            passes = field_value in val if isinstance(val, list) else field_value == val
        elif op == "not_in":
            passes = field_value not in val if isinstance(val, list) else field_value != val
        elif op == "between":
            lo, hi = val
            passes = float(lo) <= float(field_value) <= float(hi)
        else:
            passes = True
        if not passes:
            return {"decision": "blocked", "code": "condition_failed", "reason": f"Condition not met: {cond['field']} {op} {val}"}
    allowed_hours = permit.get("allowedHoursUtc") or permit.get("allowed_hours_utc")
    if isinstance(allowed_hours, list) and len(allowed_hours) == 2:
        start_h, end_h = allowed_hours
        h = now.hour
        if not (start_h <= h < end_h):
            return {"decision": "blocked", "code": "time_restricted", "reason": f"Action only allowed {start_h}:00–{end_h}:00 UTC"}
    budget_cap = float(permit.get("budgetCap", 0) or permit.get("budget_cap", 0))
    spent = float(permit.get("spent", 0) or 0)
    if spent + amount > budget_cap:
        return {"decision": "escalated", "code": "budget_exceeded", "reason": f"Action would exceed budget cap of {budget_cap}"}
    return {
        "decision": "authorized",
        "code": "ok",
        "reason": "Permit conditions satisfied",
        "spent": int(spent),
        "remaining": int(budget_cap - spent - amount),
    }

File: packages/core/policy.mjs
import crypto from 'node:crypto';

export function id(prefix = 'id') {
  return `${prefix}_${crypto.randomUUID().replaceAll('-', '').slice(0, 20)}`;
}

export function nowIso(clock = Date) { return new clock().toISOString(); }

function normalizeScope(s = '') { return String(s).trim().toLowerCase(); }

export function scopeAllows(granted, requested) {
  const g = normalizeScope(granted);
  const r = normalizeScope(requested);
  if (g === '*' || g === r) return true;
  if (g.endsWith(':*')) return r.startsWith(g.slice(0, -1));
  return false;
}

export function evaluatePermit({ permit, request, receipts = [], now = new Date() }) {
  if (!permit) return deny('permit_not_found', 'No permit was supplied');
  if (permit.status !== 'active') return deny(`permit_${permit.status}`, `Permit is ${permit.status}`);
  if (new Date(permit.expiresAt).getTime() <= now.getTime()) return deny('permit_expired', 'Permit has expired');
  if (permit.agentId !== request.agentId) return deny('agent_mismatch', 'Permit belongs to another agent');

  const allowed = (permit.scopes || []).some(s => scopeAllows(s, request.scope));
  if (!allowed) return deny('scope_denied', `Scope ${request.scope} is not granted`);

  const amount = Number(request.amount || 0);
  if (!Number.isFinite(amount) || amount < 0) return deny('invalid_amount', 'Amount must be a non-negative number');
  if (permit.maxPerAction != null && amount > Number(permit.maxPerAction)) {
    return escalate('per_action_limit', `Amount exceeds per-action limit of ${permit.maxPerAction}`);
  }

  const spent = receipts.filter(r => r.permitId === permit.id && r.result === 'authorized')
    .reduce((sum, r) => sum + Number(r.amount || 0), 0);
  if (spent + amount > Number(permit.budgetCap || 0)) {
    return escalate('budget_exceeded', `Action would exceed budget cap of ${permit.budgetCap}`);
  }

  if (permit.allowedTargets?.length && request.target && !permit.allowedTargets.includes(request.target)) {
    return deny('target_denied', 'Target is not on the permit allowlist');
  }

  // Evaluate policy conditions
  for (const cond of (permit.conditions || [])) {
    const fieldValue = request[cond.field] !== undefined ? request[cond.field] : request.context?.[cond.field];
    if (fieldValue === undefined) continue;
    let passes = false;
    switch (cond.operator) {
      case 'eq':      passes = String(fieldValue) === String(cond.value); break;
      case 'neq':     passes = String(fieldValue) !== String(cond.value); break;
      case 'lte':     passes = Number(fieldValue) <= Number(cond.value); break;
      case 'lt':      passes = Number(fieldValue) < Number(cond.value); break;
      case 'gte':     passes = Number(fieldValue) >= Number(cond.value); break;
      case 'gt':      passes = Number(fieldValue) > Number(cond.value); break;
      case 'in':      passes = Array.isArray(cond.value) ? cond.value.includes(fieldValue) : fieldValue === cond.value; break;
      case 'not_in':  passes = Array.isArray(cond.value) ? !cond.value.includes(fieldValue) : fieldValue !== cond.value; break;
      case 'between': { const [lo, hi] = cond.value; passes = Number(fieldValue) >= Number(lo) && Number(fieldValue) <= Number(hi); break; }
      default: passes = true;
    }
    if (!passes) return deny('condition_failed', `Condition not met: ${cond.field} ${cond.operator} ${JSON.stringify(cond.value)}`);
  }

  // Time-of-day restriction (allowedHoursUtc: [startHour, endHour])
  if (Array.isArray(permit.allowedHoursUtc) && permit.allowedHoursUtc.length === 2) {
    const [startH, endH] = permit.allowedHoursUtc;
    const h = now.getUTCHours();
    if (!(h >= startH && h < endH)) {
      return deny('time_restricted', `Action only allowed ${startH}:00–${endH}:00 UTC (now: ${h}:00 UTC)`);
    }
  }

  if (permit.requireHumanAbove != null && amount > Number(permit.requireHumanAbove)) {
    return escalate('human_threshold', `Amount requires human approval above ${permit.requireHumanAbove}`);
  }

  const windowMs = 60_000;
  const recent = receipts.filter(r => r.permitId === permit.id && (now.getTime() - new Date(r.createdAt).getTime()) < windowMs);
  if (permit.rateLimitPerMinute && recent.length >= permit.rateLimitPerMinute) {
    return deny('rate_limited', 'Permit rate limit reached');
  }

  return { decision: 'authorized', code: 'ok', reason: 'Permit conditions satisfied', spent, remaining: Number(permit.budgetCap || 0) - spent - amount };
}

function deny(code, reason) { return { decision: 'blocked', code, reason }; }
function escalate(code, reason) { return { decision: 'escalated', code, reason }; }
